flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java

```diff
@@ -1480,4 +1513,27 @@ private void checkIdleSlotTimeout() {
                         settings.getSlotIdleTimeout().toMillis(),
                         TimeUnit.MILLISECONDS);
     }
+
+    private CheckpointStatsListener createCheckpointStatsListener() {
+        // wrapper implementation ensuring that the checkpoint-related events are handled in the
+        // scheduler's main thread
```

Comment:
This assumption doesn't seem to be valid. The `state.tryRun(` doesn't schedule an action on the main thread.

Reply:
Yikes, good catch. 😱  We're not ensure that the events are handled in the main thread. I will updated that. 👍